### DISCUSSION TITLE
Understanding renovate and used docker digests || wrong docker digests?

### DISCUSSION THREAD

**Discussion by druesendieb** [renovatebot/renovate#11391] [Request Help] [answered]
Hi folks,

we're just building a PoC with renovate to easier cope with technical debt and from what I've seen so far: really awesome product, thanks for your great work!

We're using the self-hosted variant in Gitlab, which also hosts our container registry. When updating the first docker images (using default config) I stumbled upon some unexpected behaviour that got me quite confused and I want to understand.

The digests of the new images in the `DOCKERFILE` doesn't correspond with the current digests on docker hub.
I've tried 2 versions of renovate, `v24.x` and `v.26.3.0`

So when onboarding the imagerepo `hashicorp/vault-k8s`  I got the following preview in the MR:

<details>
<summary>Update hashicorp/vault-k8s Docker tag to v0.12.0</summary>

  - Schedule: ["at any time"]
  - Branch name: `renovate/docker-hashicorp-vault-k8s-0.x`
  - Merge into: `master`
  - Upgrade hashicorp/vault-k8s to `sha256:ee6330531f4d75ad3a38e038fb04486353f1d6e19798e72bf84a6922fde54160`

</details>

But when I look on docker hub, I can't find an image corresponding with the digest et al. (I thought about maybe being an os/arch problem)
![digestconfusion](https://user-images.githubusercontent.com/2674011/130488183-f41d242c-6a59-428f-9be1-0e9800eb18eb.png)

Expected behaviour was to get the following new digest:
`hashicorp/vault-k8s:0.12.0@sha256:81c09f6d42c2db8121bcd759565ea244cedc759f36a0f090ec7da9de4f7f8fe4` 
![digestconfusion1](https://user-images.githubusercontent.com/2674011/130488585-74982c3d-0177-46d6-92a0-5c7384e547e3.png)

I thought, maybe there's an issue with the preview in the onboarding MR and onboarded the repository, but the new MR for updating the images contain the same digests.

Can please somebody shed some light on this?

**Comment by viceice** [comment, marked as answer]
Renovate is using the manifest list to which the `0.12.0` tag points.

```sh
docker manifest inspect hashicorp/vault-k8s:0.12.0
docker pull hashicorp/vault-k8s:0.12.0
docker inspect hashicorp/vault-k8s:0.12.0
```
```json
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1156,
         "digest": "sha256:81c09f6d42c2db8121bcd759565ea244cedc759f36a0f090ec7da9de4f7f8fe4",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1156,
         "digest": "sha256:2ffd023131d0184923d237f18f4ad5009b8afe71d83ded7b4c272db204ab7b8a",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1156,
         "digest": "sha256:295b5db1283633daf8d82f0b794bd543118508a5568c90b055d7a48ad6afcfd1",
         "platform": {
            "architecture": "386",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1156,
         "digest": "sha256:197c58a9c30faf97f16412d63e333a0adffed856b06994759899d070f28db51e",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v6"
         }
      }
   ]
}
```

```json
[
    {
        "Id": "sha256:a5d44c0fc90734573afaa286e5138f14d3a9a067493052fa8f7cba950301b09b",
        "RepoTags": [
            "hashicorp/vault-k8s:0.12.0"
        ],
        "RepoDigests": [
            "hashicorp/vault-k8s@sha256:ee6330531f4d75ad3a38e038fb04486353f1d6e19798e72bf84a6922fde54160"
        ],
        "Parent": "",
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2021-08-18T15:57:50.017113203Z",
        "Container": "",
        "ContainerConfig": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "DockerVersion": "",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "vault",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "VERSION=0.12.0",
                "LOCATION=https://releases.hashicorp.com"
            ],
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/bin/vault-k8s"
            ],
            "OnBuild": null,
            "Labels": {
                "description": "Vault-K8s includes first-class integrations between Vault and Kuberentes. Integrations include the Vault Agent Injector mutating admission webhook.",
                "maintainer": "Vault Team <vault@hashicorp.com>",
                "name": "Vault K8s",
                "release": "0.12.0",
                "summary": "The Vault-K8s binary includes first-class integrations between Vault and Kubernetes.",
                "vendor": "HashiCorp",
                "version": "0.12.0"
            }
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 54228400,
        "VirtualSize": 54228400,
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/86f2ddb8d8050f5bb396a029a265a07941289e25125837d94a07ce82435db5e3/diff:/var/lib/docker/overlay2/38ba725996083ba973afbcd4934bb308ba4fc6881be935ec9d260c143173cd67/diff:/var/lib/docker/overlay2/3f8a05a0e62f952b1618ec0338ac0d2cf2e77139219998d0767fe702e05f1ee6/diff",
                "MergedDir": "/var/lib/docker/overlay2/3901a00eecce7b5fbe38eec6168d2a90262932d29c097be6373e43993bd8bf69/merged",
                "UpperDir": "/var/lib/docker/overlay2/3901a00eecce7b5fbe38eec6168d2a90262932d29c097be6373e43993bd8bf69/diff",
                "WorkDir": "/var/lib/docker/overlay2/3901a00eecce7b5fbe38eec6168d2a90262932d29c097be6373e43993bd8bf69/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:bc276c40b172b1c5467277d36db5308a203a48262d5f278766cf083947d05466",
                "sha256:ffe6e4a3ab02f97e6506ee63b6f3a22f95bc412a473bfe7ad54c1d915b1b90d1",
                "sha256:a595e7ed23c5a05bc8532d2af4f17f8f2fc692424c46f1268fc130405c403a4c",
                "sha256:e05b172482514f12da83bdf2d5d082e9573574bfe8523d155263584e44af30f5"
            ]
        },
        "Metadata": {
            "LastTagTime": "0001-01-01T00:00:00Z"
        }
    }
]
```

So yes, that digest is expected

**Reply by druesendieb** [reply]
Thanks for your answer so far, but actually this confuses me even more, so here are some follow-up questions I was unlucky resolving by googling 😕 

If renovate uses the manifest list, how is it possible to get the RepoDigest if it's not in the manifest list?
For me as a human, who wants to double-check digests for security reasons - is it possible to view the RepoDigest e.g. on Dockerhub? 

Also, when pulled with the os/arch-specific digest, in the end I get 2 RepoDigests for (what i thoguth should be) the same image but I don't understand quite why.



```
# docker pull hashicorp/vault-k8s:0.12.0
# docker pull hashicorp/vault-k8s:0.12.0@sha256:81c09f6d42c2db8121bcd759565ea244cedc759f36a0f090ec7da9de4f7f8fe4
# docker inspect hashicorp/vault-k8s:0.12.0
[
    {
        "Id": "sha256:a5d44c0fc90734573afaa286e5138f14d3a9a067493052fa8f7cba950301b09b",
        "RepoTags": [
            "hashicorp/vault-k8s:0.12.0"
        ],
        "RepoDigests": [
            "hashicorp/vault-k8s@sha256:81c09f6d42c2db8121bcd759565ea244cedc759f36a0f090ec7da9de4f7f8fe4",
            "hashicorp/vault-k8s@sha256:ee6330531f4d75ad3a38e038fb04486353f1d6e19798e72bf84a6922fde54160"
        ],
]
(...)

# docker images --digests | grep vault-k8s
hashicorp/vault-k8s                                                   0.12.0                sha256:81c09f6d42c2db8121bcd759565ea244cedc759f36a0f090ec7da9de4f7f8fe4   a5d44c0fc907        5 days ago          54.2MB
hashicorp/vault-k8s                                                   0.12.0                sha256:ee6330531f4d75ad3a38e038fb04486353f1d6e19798e72bf84a6922fde54160   a5d44c0fc907        5 days ago          54.2MB

**Reply by viceice** [reply]
Renovate behaves like the docker client, it fetches the manifest ( can be a single docker manifest or a manifest list).

If the result is a manifest list, renovate uses the manifest list digest as update otherwise it uses the image manifest digest.

Sadly docker hub don't show the manifest list digest, you only get it when you pull it via docker client (see my commands above).

So `sha256:81c09f6d42c2db8121bcd759565ea244cedc759f36a0f090ec7da9de4f7f8fe4` is the digest of the amd64 image manifest and `sha256:ee6330531f4d75ad3a38e038fb04486353f1d6e19798e72bf84a6922fde54160` is the digest of the manifest list.

You can't modify a manifest list without creating a new digest. so it's enough to verify the manifestlist digest.

https://docs.docker.com/registry/spec/manifest-v2-2/


@rarkins As we pass now the current value and the current digest, we can check if the current digest is pointing to a manifest list or an image manifest. So we could propose a update to the real image if the current digest is a image manifest. WDYT ? I'll open an issue for it

**Reply by viceice** [reply]
This can also fix the bazel issue which expects the amd64 digests, #9786

**Reply by druesendieb** [reply]
Aaaah, the digest is of the manifest list!
Thanks a lot for the clarifications, this makes sense now.

From a user perspective, we would love to use the image manifest as this enables pinning an architecture and cross-checking image digests without pulling first!

**Reply by rarkins** [reply]
@viceice let's have an issue to discuss the idea. I'm not sure how easy it will be to use the existing digest to reliably determine which "new" digest to choose.. certainly in O(1) time

**Comment by viceice** [comment]
#11409

**Comment by marcofranssen** [comment]
To get the digest of any image you can use the tool crane

```
crane digest myimage
```

crane has many other useful utilities to get information from images.
